libp2p/security/insecure/transport.py

```diff
@@ -1,15 +1,55 @@
+from libp2p.crypto.keys import PublicKey
+from libp2p.crypto.pb import crypto_pb2
+from libp2p.crypto.utils import pubkey_from_protobuf
 from libp2p.network.connection.raw_connection_interface import IRawConnection
 from libp2p.peer.id import ID
 from libp2p.security.base_session import BaseSession
 from libp2p.security.base_transport import BaseSecureTransport
 from libp2p.security.secure_conn_interface import ISecureConn
+from libp2p.transport.exceptions import HandshakeFailure
 from libp2p.typing import TProtocol
+from libp2p.utils import encode_fixedint_prefixed, read_fixedint_prefixed
 
-PLAINTEXT_PROTOCOL_ID = TProtocol("/plaintext/1.0.0")
+from .pb import plaintext_pb2
+
+# Reference: https://github.com/libp2p/go-libp2p-core/blob/master/sec/insecure/insecure.go
+
+
+PLAINTEXT_PROTOCOL_ID = TProtocol("/plaintext/2.0.0")
 
 
 class InsecureSession(BaseSession):
-    pass
+    async def run_handshake(self):
+        msg = make_exchange_message(self.local_private_key.get_public_key())
+        msg_bytes = msg.SerializeToString()
+        encoded_msg_bytes = encode_fixedint_prefixed(msg_bytes)
+        await self.write(encoded_msg_bytes)
+
+        remote_msg_bytes = await read_fixedint_prefixed(self.conn)
+        remote_msg = plaintext_pb2.Exchange()
+        remote_msg.ParseFromString(remote_msg_bytes)
+
+        # Verify if the given `pubkey` matches the given `peer_id`
+        try:
+            remote_pubkey = pubkey_from_protobuf(remote_msg.pubkey)
+        except ValueError:
+            raise HandshakeFailure(
+                f"unknown `key_type` of remote_msg.pubkey={remote_msg.pubkey}"
+            )
+        remote_peer_id = ID(remote_msg.id)
+        remote_peer_id_from_pubkey = ID.from_pubkey(remote_pubkey)
+        if remote_peer_id_from_pubkey != remote_peer_id:
+            raise HandshakeFailure(
+                "peer id and pubkey from the remote mismatch: "
+                f"remote_peer_id={remote_peer_id}, remote_pubkey={remote_pubkey}, "
+                f"remote_peer_id_from_pubkey={remote_peer_id_from_pubkey}"
+            )
+
+        # Nothing is wrong. Store the `pubkey` and `peer_id` in the session.
+        self.remote_permanent_pubkey = remote_pubkey
+        self.remote_peer_id = ID(remote_msg.id)
```

Comment:
`self.remote_peer_id` is initialized during `__init__`, should we also compare `remote_peer_id` against `self.remote_peer_id` before setting it to new value?
https://github.com/libp2p/py-libp2p/blob/6b05d9ab8f0de5e15f8a65ffaa199e760d209cbe/libp2p/security/base_session.py#L16-L23

Reply:
Good point. Added the check in the latest commit.